Fix IItemHandler wrappers returned by EntityLivingBase

Choonster opened the pull request:
`EntityLivingBase#getCapability` returns `IItemHandler` wrappers of `EntityLivingBase#handInventory` (`field_184630_bs`) and `EntityLivingBase#armorArray` (`field_184631_bt`), but the contents of these lists are replaced every tick in `EntityLivingBase#onUpdate` by the return values of `EntityLivingBase#getItemStackFromSlot`. This means that any changes made through the `IItemHandler`s are lost next tick.

This PR adds new `IItemHandler` wrapper classes that call `EntityLivingBase#getItemStackFromSlot`/`EntityLivingBase#setItemStackToSlot` instead of using the item lists directly. These wrappers are returned by `EntityLivingBase#getCapability`.

[This test mod](https://github.com/Choonster-Minecraft-Mods/EntityLivingBaseItemHandlerTestMod) can be used to reproduce the issue:
* Right click on an entity with a stick to open an inventory GUI for it
* Try to place items in the entity's equipment slots:
    * If it's a vanilla entity in a regular build of Forge, the items will disappear after a moment.
    * If it's a vanilla entity in a build of Forge with this fix applied, the items will persist as expected.
    * If it's an ItemHandler Fix Zombie (added by the mod), the items will persist as expected even without this fix applied.

`EntityPlayer` is the only direct child of `EntityLivingBase` not affected by this, since it already has its own override of `EntityLivingBase#getCapability`.

~~The behaviour of the `IItemHandler`s returned by `EntityLivingBase#getCapability` isn't obvious and isn't documented anywhere, should this method be removed/deprecated or should the `IItemHandler`s be replaced with read-only wrappers that reject modifications instead of accepting and then silently discarding them?~~

Edit: Update to mention the new wrappers instead of the previous overrides.

Choonster committed:
Fix IItemHandler wrappers returned by EntityLiving/EntityArmorStand

williewillus commented:
I added the patches, didn't know the arrays were badly named.
The intent is that you are able to see the armor/hands of any entity as an inventory and read/write at will.

LexManos commented:
Probably the better solution would be to write wrappers that bounce any set calls to setItemStackToSlot.

Choonster committed:
Replace EntityLivingBase IItemHandlers with custom wrappers

- Remove EntityLivingBase#getCapability overrides from EntityLiving and EntityArmorStand

Choonster commented:
I've created wrapper classes and removed the `EntityLivingBase#getCapability` overrides from `EntityLiving` and `EntityArmorStand`, as per your suggestion.

ichttt commented on src/main/java/net/minecraftforge/items/wrapper/EntityEquipmentInvWrapper.java:
```diff
@@ -0,0 +1,170 @@
+    protected EntityEquipmentSlot validateSlotIndex(final int slot)
+    {
+        if (slot < 0 || slot >= slots.size())
+            throw new RuntimeException("Slot " + slot + " not in valid range - [0," + slots.size() + ")");
```
I think an IndexOutOfBoundsException would be better here

bs2609 replied:
An `IllegalArgumentException` might make more sense, there doesn't seem to be much point in throwing `IndexOutOfBoundsException` here -  `List.get()` will throw it for the exact same condition.

Choonster replied:
I just copied this method from `ItemStackHandler`, but throwing `IllegalArgumentException` makes sense.

Choonster committed:
Throw IllegalArgumentException instead of RuntimeException

LexManos commented on src/main/java/net/minecraftforge/items/wrapper/EntityEquipmentInvWrapper.java:
```diff
@@ -0,0 +1,170 @@
+            {
+                slots.add(slot.getIndex(), slot);
+            }
+        }
```
Not a fan of this, it would seem to add non-congruent slots. So the inventory would appear to have 'holes'
A better way may be using add instead of a indexed add.

Choonster replied:
Unless someone adds a new `EntityEquipmentSlot` at runtime, this will never have holes. Vanilla uses the same values as the indexes of its item lists, so this guarantees that the `IItemHandler` uses the same indexes.

The enum constants *are* ordered by their index, so it would produce the same result.

Alpvax commented:
> Unless someone adds a new EntityEquipmentSlot at runtime

But that isn't necessarily as unlikely as it would first seem. This is modded minecraft.

Choonster commented:
> But that isn't necessarily as unlikely as it would first seem. This is modded minecraft.

Anyone doing that would also be responsible for ensuring that the index of the new value is unique and contiguous with the existing values; but it may be best to just rely on the enum's natural ordering rather than manually-assigned indexes (as Lex suggested).

Choonster committed:
Use the natural ordering of the EntityEquipmentSlot values instead of the manually-assigned indexes

Choonster commented:
I just noticed a minor error in the documentation: The doc comment for `EntityEquipmentInvWrapper#slots` still says that `EntityEquipmentSlot#index` is used as the index, but this is no longer the case.

Sorry for only bringing this up after you merged the PR, but I missed it earlier.

I'm guessing you probably don't want a new PR for a one-line documentation change.

kashike commented:
You forgot the licence headers, as well.
